lib/awful/placement.lua

```diff
@@ -155,6 +159,89 @@ local function move_into_geometry(source, target)
     return ret
 end
 
+-- Update the workarea
+local function wibox_update_strut(d, position)
+    -- If the drawable isn't visible, remove the struts
+    if not d.visible then
+        d:struts { left = 0, right = 0, bottom = 0, top = 0 }
+        return
+    end
+
+    -- Detect horizontal or vertical drawables
+    local geo      = area_common(d)
+    local vertical = geo.width < geo.height
+
+    -- Look into the `position` string to find the relevants sides to crop from
+    -- the workarea
+    local struts = { left = 0, right = 0, bottom = 0, top = 0 }
+
+    if vertical then
+        for _, v in ipairs {"right", "left"} do
+            if (not position) or position:match(v) then
+                struts[v] = geo.width
+            end
+        end
+    else
+        for _, v in ipairs {"top", "bottom"} do
+            if (not position) or position:match(v) then
+                struts[v] = geo.height
+            end
+        end
+    end
+
+    -- Update the workarea
+    d:struts(struts)
+end
+
+--- Pin a drawable to a placement function.
+-- Automatically update the position when the size change.
+-- All other arguments will be passed to the `position` function (if any)
+-- @tparam[opt=client.focus] drawable d A drawable (like `client`, `mouse`
+--   or `wibox`)
+-- @param position_f A position name (see `align`) or a position function
+-- @tparam[opt={}] table args Other arguments
+local function attach(d, position_f, args)
+    args = args or {}
+
+    if not args.attach then return end
+
+    d = d or capi.client.focus
+    if not d then return end
+
+    if type(position_f) == "string" then
+        position_f = placement[position_f]
+    end
+
+    if not position_f then return end
+
+    local function tracker()
+        position_f(d, args)
+    end
+
+    d:connect_signal("property::width" , tracker)
+    d:connect_signal("property::height", tracker)
+
+    tracker()
+
+    if args.update_workarea then
+        local function tracker_struts()
+            --TODO this is too fragile and doesn't work with all methods.
+            wibox_update_strut(d, reverse_align_map[position_f])
```

Comment:
Oops, I forgot a few lines when building this commit. The align part is not test-covered yet. I will have to write a rest `tests/test-*.lua` for this as my testing framework lack the code to emulate workarea (and it would be a PITA to implement)